rfcs/010-gateway-refactor/010-gateway-refactor.md

```diff
@@ -0,0 +1,205 @@
+looks this ID up in redis and if there is a match then any tokens/credentials stored
+under that key can be added to requests initiated by that ID.
+
+These sessions are managed/started as follows:
+- when a user logs in they get a session
+- when a user server is started the server gets a redis-based session
+- when a login happens through the CLI there is a also a session
+- anonymous users also have redis-based sessions but these are empty
+since there is no additional credentials that need to be managed for anonymous users
+
+The gateway then has the following responsibilities:
+- store refresh and access tokens
+- keep all tokens stored up to date and manage refreshing tokens
+- destroy sessions when certain events occur (i.e. when a logout occurs)
+
+### Centralized Jupyter server authentication
+
+### Diagrams
+
+1. Creating a user server
+
+```mermaid
+graph TD
+    Ingress -- 1<br>Session ID --> Gateway
+    
+    subgraph GatewayGroup[ ]
+    Gateway
+    DB[(DB)]
+    Gateway -- 2<br>Session ID --> DB
+    Gateway -- 3<br>Server ID --> DB
+    DB -- 4<br>KC access token<br>Server ID --> Gateway
+    end    
+    
+    Gateway -- 5<br>KC access token<br>Server ID--> Notebook[Notebook service]
+    Notebook -- 6<br>KC user ID<br>Server ID--> Server[User server]
+```
+
+2. Accessing a user server
+
+```mermaid
+graph TD
+    Ingress -- 1<br>Session ID --> Gateway
+    
+    subgraph GatewayGroup[ ]
+    Gateway
+    DB[(DB)]
+    Gateway -- 2<br>Session ID --> DB
+    DB -- 3<br>KC access token --> Gateway
+    end    
+    
+    Gateway -- 4<br>KC access token --> Proxy
+
+    subgraph Server[User server]
+    Proxy -- 5 --> Jupyter
+    end
+```
+
+3. User server sidedcar accessing a protected service
+
+```mermaid
+graph TD
+    Server[User server sidecar]
+    Gitlab
+
+    subgraph GatewayGroup[ ]
+    Gateway
+    DB[(DB)]
+    end  
+
+    Server -- 1<br>Server ID --> Gateway
+    Gateway -- 2<br>Server ID --> DB
+    DB -- 3<br>Git access token --> Gateway
+    Gateway -- 4<br>Git access token --> Gitlab
+```
+
+4. Sticky sessions for a Renku service
+
+```mermaid
+graph TD
+    Ingress
+    
+    subgraph CoreGroup[ ]
+        CoreTraefik[Traefik]
+        Core[Core service]
+    end
+   
+    subgraph GatewayGroup[ ]
+        Gateway
+        DB[(DB)]
+    end  
+    
+    Ingress -- 1<br>Session ID --> CoreTraefik
+    CoreTraefik -- 2<br>Session ID<br>Sticky session --> Gateway
+    Gateway -- 3<br>Session ID --> DB
+    DB -- 4<br>KC access token<br>Git access token --> Gateway
+    Gateway -- 5<br>KC access token<br>Git access token<br>Sticky session --> Core
+```
+
+5. Requests from the ui-server
+
+Here the gateway will simply require fully valid Keycloak access tokens
+(which the ui-server already has). If this is implemented, the ui-server
+will be the only component that will be coming to the gateway with Keycloak
+access tokens. All others will be coming with a session ID. So to further
```

Comment:
seems like the ui-server should conform to this architecture instead?

Reply:
Yes but the ui-server already has sessions with each user and it keeps access tokens for these sessions up to date. Removing this functionality from the ui-server and merging it with the gateway was too big of a change IMO. And it is something that I am not sure whether we should or should not do. So accepting KC access tokens in the gateway  was the simplest thing to do. 

The alternative to this is getting the ui-server to have a session with the gateway. But there would need to be one such session per user. Not just one for the whole service. I think this definitely complicates things and makes it so that we need changes in the ui-server in addition to the gateway.

Reply:
I agree that the UI server should conform: it should be treated as a client like the others (_there might be a few complications so I'm happy if we postpone it to phase 2 and initially accept valid access tokens_).

We can handle it by keeping the session with the client, and the server will map that to another session with the gateway. This way, we will just attach/swap the session cookies to every request instead of swapping with the user token.
Ideally, the change is not that big on our side.